modules/index-rebuild/src/main/java/org/opencastproject/index/rebuild/IndexRebuildService.java

```diff
@@ -0,0 +1,228 @@
+    ServiceReference<?>[] serviceReferences = bundleContext.getAllServiceReferences(IndexProducer.class.getName(),
+            null);
+    if (serviceReferences != null) {
+      for (ServiceReference<?> serviceReference : serviceReferences) {
+        addIndexProducer((IndexProducer) bundleContext.getService(serviceReference));
+      }
+    }
+
+    // all available?
+    if (indexProducers.size() == IndexRebuildService.Service.values().length) {
+      registerIndexRebuildService(bundleContext);
+    }
+
+    // listen to changes in availability
+    bundleContext.addServiceListener(new IndexProducerListener(bundleContext),
+            "(objectClass=" + IndexProducer.class.getName() + ")");
+  }
+
+  @Override
+  public void stop(BundleContext bundleContext) throws Exception {
+    unregisterIndexRebuildService();
+  }
+
+  /**
+   * Recreate the index from all of the services that provide data for it.
+   *
+   * @param index
+   *           The index to rebuild.
+   *
+   * @throws IOException
+   *           Thrown if the index cannot be cleared.
+   * @throws IndexRebuildException
+   *           Thrown if the index rebuild failed.
+   */
+  public synchronized void recreateIndex(AbstractSearchIndex index)
+          throws IOException, IndexRebuildException {
+    index.clear();
+    for (IndexRebuildService.Service service: IndexRebuildService.Service.values()) {
+      recreateService(index, service);
+    }
+  }
+
+  /**
+   * Recreate the index from a specific service that provides data for it.
+   *
+   * @param index
+   *           The index to rebuild.
+   * @param serviceName
+   *           The name of the {@link Service}.
+   *
+   * @throws IllegalArgumentException
+   *           Thrown if the service doesn't exist.
+   * @throws IndexRebuildException
+   *           Thrown if the index rebuild failed.
+   */
+  public synchronized void recreateIndex(AbstractSearchIndex index, String serviceName)
+          throws IllegalArgumentException, IndexRebuildException {
+    IndexRebuildService.Service service = IndexRebuildService.Service.valueOf(serviceName);
+    recreateService(index, service);
+  }
+
+  /**
+   * Trigger repopulation of the specified index with data from the specified service.
+   *
+   * @param index
+   *           The index to rebuild.
+   * @param service
+   *          The {@link IndexRebuildService.Service} to re-add data from.
+   *
+   * @throws IndexRebuildException
+   *           Thrown if the index rebuild failed.
+   */
+  private void recreateService(AbstractSearchIndex index, IndexRebuildService.Service service)
+          throws IndexRebuildException {
+
+    if (!indexProducers.containsKey(service)) {
+      throw new IllegalStateException(format("Service %s is not available", service));
+    }
+
+    IndexProducer indexProducer = indexProducers.get(service);
+    logger.info("Starting to recreate index {} for service '{}'", index.getIndexName(), service);
```

Comment:
Same here. This will not recreate the index, will it? The index is already there and existing entries would not be modified, would they?

Reply:
See above. Well, in this instance, the index might have been cleared before, but we don't really know that. I rewrote so many of those senseless log messages, but I still missed some...